tools/ioflow/ioflow-read.c

```diff
@@ -0,0 +1,225 @@
+/* 
+ * ioflow-read.c        C source code for read tracer. 
+ *
+ *
+ * Copyright (c) Google LLC
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ *
+ * 25-Aug-2020   Haoning Chen   Created this.
+ */
+
+#include <uapi/linux/ptrace.h>
+#include <linux/blkdev.h>
+#include <linux/fs.h>
+
+#define IO_FLOW_READ 1
+// Use python string replacement to import common header code
+[IMPORT_COMM]
+
+BPF_HASH(map, u64, struct val_t);    // Map of syscall data, with pid as key
+BPF_HASH(rqmap, struct request *, struct rqval_t);    // Map of request info for aync request handling
+
+BPF_HISTOGRAM(hist_vfs);
+BPF_HISTOGRAM(hist_pgcache);
+BPF_HISTOGRAM(hist_readpg);
+BPF_HISTOGRAM(hist_ext4readpg);
+BPF_HISTOGRAM(hist_blk);
+BPF_HISTOGRAM(hist_split);
+BPF_HISTOGRAM(hist_merge);
+// Async request handling histograms
+BPF_HISTOGRAM(hist_rq_queue);
+BPF_HISTOGRAM(hist_rq_service);
+
+BPF_PERF_OUTPUT(syscall_events);
+BPF_PERF_OUTPUT(rq_events);
+
+
+int vfs_read_entry(struct pt_regs *ctx, struct file *file) {
+    // Filter asyc and direct IO
+    if (!(file->f_op->read_iter) || file->f_flags & O_DIRECT)
```

Comment:
Would it be possible for us to trace O_DIRECT too?

Reply:
I didn't look into function call stack for direct IO due to time constraints, but our current code can definitely trace dio, since it also reaches the block layer.